File: swift/Sources/AetherMeshProtocol/Transport/PredictiveTransportSelector.swift
// SPDX-License-Identifier: MIT
// Predictive transport selector — 2-state Kalman RTT filter over PerTransportMetrics.
//
// Why Kalman over EWMA?
// ─────────────────────
// EWMA is a 1-pole IIR: it smooths past measurements but cannot predict future RTT
// when a link is actively degrading.  The Kalman filter models RTT as a constant-
// velocity process [rtt, drift]:
//
//   x_t = F × x_{t−1} + w   (F = [[1,1],[0,1]])
//   z_t = H × x_t   + v    (H = [1,0])
//
// Positive drift signals a rising RTT *before* it exceeds a threshold, enabling
// proactive transport switching.  The posterior variance further penalises uncertain
// links even when their point estimate looks good.
//
// Score formula:
//   (effectiveBps / powerCost) × (1 − lossRate) / max(kalmanRtt, 1) × (1 / (1 + σ/100))
//
// Thread-safe via NSLock.
//
// NOTE: TransportService is AnyObject, so ObjectIdentifier is used as dictionary key
// to enable pointer-identity comparisons without requiring Hashable conformance.

import Foundation

// ── KalmanRttFilter ───────────────────────────────────────────────────────────

/// Two-state Kalman filter estimating RTT and drift for a single transport link.
///
/// State: x = [rtt; drift] — F = [[1,1],[0,1]], H = [1,0].
///
/// **Not thread-safe.** Callers must hold the selector's lock before calling `update`.
private final class KalmanRttFilter {

    private let qRtt:   Double
    private let qDrift: Double
    private let r:      Double

    private(set) var rtt:   Double
    private(set) var drift: Double
    private(set) var p00:   Double
    private(set) var p01:   Double
    private(set) var p11:   Double

    /// Posterior variance of the RTT estimate (ms²). Lower = more confident.
    var rttVariance: Double { p00 }

    init(
        initialRttMs: Double = 200.0,
        qRtt:         Double = 25.0,
        qDrift:       Double = 5.0,
        r:            Double = 100.0
    ) {
        self.rtt    = initialRttMs
        self.drift  = 0.0
        self.p00    = 400.0
        self.p01    = 0.0
        self.p11    = 100.0
        self.qRtt   = qRtt
        self.qDrift = qDrift
        self.r      = r
    }

    /// Incorporate a new RTT measurement and return the updated estimate.
    @discardableResult
    func update(measuredRttMs: Double) -> Double {
        // ── 1. Predict ────────────────────────────────────────────────────────
        let rttPred   = rtt + drift
        let driftPred = drift

        // P_pred = F·P·Fᵀ + Q  (F = [[1,1],[0,1]])
        let pp00 = p00 + 2.0 * p01 + p11 + qRtt
        let pp01 = p01 + p11
        let pp11 = p11 + qDrift

        // ── 2. Kalman gain (H = [1, 0]) ──────────────────────────────────────
        let S  = pp00 + r
        let k0 = pp00 / S
        let k1 = pp01 / S

        // ── 3. Update ─────────────────────────────────────────────────────────
        let innovation = measuredRttMs - rttPred
        rtt   = rttPred   + k0 * innovation
        drift = driftPred + k1 * innovation

        // P = (I − K·H)·P_pred
        p00 = (1.0 - k0) * pp00
        p01 = (1.0 - k0) * pp01
        p11 = -k1 * pp01 + pp11

        // Clamp to prevent numerical drift below zero.
        p00 = max(p00, 1e-6)
        p11 = max(p11, 1e-6)

        return rtt
    }
}

// ── PredictiveTransportSelector ───────────────────────────────────────────────

/// A transport paired with its Kalman-predictive score and uncertainty metadata.
public struct PredictedRankedTransport {
    /// The ranked transport backend.
    public let transport:      any TransportService
    /// Composite predictive score (higher = better).
    public let score:          Double
    /// Kalman-estimated RTT in milliseconds.
    public let predictedRttMs: Double
    /// Posterior RTT variance (ms²). Lower = more confident.
    public let rttVariance:    Double
}

/// Predictive transport selector maintaining a per-transport Kalman RTT filter.
///
/// `TransportService` conforms to `AnyObject`, so `ObjectIdentifier` is used as
/// the dictionary key for stable pointer-identity comparisons without requiring
/// `Hashable` conformance on the protocol.
///
/// Thread-safe via `NSLock`.
public final class PredictiveTransportSelector: @unchecked Sendable {

    private let lock = NSLock()

    // Keyed by ObjectIdentifier so we don't need TransportService: Hashable.
    private var filters:    [ObjectIdentifier: KalmanRttFilter]       = [:]
    private var transports: [ObjectIdentifier: any TransportService]  = [:]

    public init() {}

    // ── Registration ──────────────────────────────────────────────────────────

    /// Register a transport for Kalman tracking with an initial RTT prior.
    /// Safe to call multiple times — subsequent calls for already-registered
    /// transports are no-ops.
    public func register(_ transport: any TransportService, initialRttMs: Double = 200.0) {
        let id = ObjectIdentifier(transport)
        lock.withLock {
            guard filters[id] == nil else { return }
            filters[id]    = KalmanRttFilter(initialRttMs: initialRttMs)
            transports[id] = transport
        }
    }

    /// Remove a transport and discard its Kalman state.
    public func unregister(_ transport: any TransportService) {
        let id = ObjectIdentifier(transport)
        lock.withLock {
            filters.removeValue(forKey: id)
            transports.removeValue(forKey: id)
        }
    }

    // ── Observation ───────────────────────────────────────────────────────────

    /// Feed a new sample to both the transport's PerTransportMetrics EWMA and
    /// our Kalman filter.  Call after every completed send attempt.
    ///
    /// - Parameters:
    ///   - transport:        The transport that just completed a send.
    ///   - rttMs:            Measured round-trip time in ms.
    ///   - success:          Whether the peer acknowledged receipt.
    ///   - bytesTransferred: Bytes successfully transferred.
    public func observeMetrics(
        _ transport:       any TransportService,
        rttMs:             Double,
        success:           Bool,
        bytesTransferred:  Int
    ) {
        transport.metrics?.recordSample(
            rttMs: rttMs, success: success, bytesTransferred: bytesTransferred)

        guard rttMs > 0, success else { return }

        let id = ObjectIdentifier(transport)
        lock.withLock {
            filters[id]?.update(measuredRttMs: rttMs)
        }
    }

    // ── Ranking ───────────────────────────────────────────────────────────────

    /// Returns transports in descending predictive-score order.
    ///
    /// Only available transports are included.  `payloadBytes` is used to
    /// exclude transports too slow to deliver this payload within 30 s.
    public func rank(payloadBytes: Int = 512) -> [PredictedRankedTransport] {
        var result: [PredictedRankedTransport] = []

        lock.withLock {
            for (id, filter) in filters {
                guard let transport = transports[id] else { continue }
                guard transport.isAvailable else { continue }

                let bw = transport.maxBandwidthBps
                if bw > 0 {
                    let serialSec = Double(payloadBytes) * 8.0 / Double(bw)
                    if serialSec > 30.0 { continue }
                }

                let kalmanRtt = max(filter.rtt, 1.0)
                let variance  = filter.rttVariance
                let stddev    = variance.squareRoot()
                let power     = Double(max(transport.powerCostRelative, 1))

                let lossRate: Double
                let effectiveBps: Double

                if let m = transport.metrics {
                    lossRate     = m.ewmaLossRate
                    effectiveBps = max(m.ewmaThroughputBps, Double(bw) * 0.1)
                } else {
                    lossRate     = 0.05
                    effectiveBps = Double(bw) * 0.1
                }

                // Reliability factor: 1.0 at σ=0 ms, ~0.5 at σ=100 ms.
                let reliabilityFactor = 1.0 / (1.0 + stddev / 100.0)
                let score = (effectiveBps / power) * (1.0 - lossRate) / kalmanRtt * reliabilityFactor

                result.append(PredictedRankedTransport(
                    transport:      transport,
                    score:          score,
                    predictedRttMs: kalmanRtt,
                    rttVariance:    variance
                ))
            }
        }

        return result.sorted { $0.score > $1.score }
    }

    /// Returns the highest-scoring available transport, or `nil`.
    public func selectBest(payloadBytes: Int = 512) -> (any TransportService)? {
        rank(payloadBytes: payloadBytes).first?.transport
    }

    /// Returns `(rttMs, driftMs, variance)` for a registered transport, or `nil`.
    public func kalmanState(
        for transport: any TransportService
    ) -> (rttMs: Double, driftMs: Double, variance: Double)? {
        let id = ObjectIdentifier(transport)
        return lock.withLock {
            guard let f = filters[id] else { return nil }
            return (f.rtt, f.drift, f.p00)
        }
    }
}

File: kotlin/src/main/kotlin/aethernet/transport/PredictiveTransportSelector.kt
// SPDX-License-Identifier: MIT
// Predictive transport selector — 2-state Kalman RTT filter over PerTransportMetrics.
//
// Why Kalman over EWMA?
// ─────────────────────
// EWMA is a 1-pole IIR: it smooths past measurements but cannot predict future RTT
// when a link is actively degrading.  The Kalman filter models RTT as a constant-
// velocity process [rtt, drift]:
//
//   x_t = F * x_{t−1} + w   (F = [[1,1],[0,1]])
//   z_t = H * x_t   + v    (H = [1,0])
//
// Positive drift signals a rising RTT *before* it exceeds a threshold, enabling
// proactive transport switching.  The posterior variance further penalises uncertain
// links even when their point estimate looks good.
//
// Score formula:
//   (effectiveBps / powerCost) × (1 − lossRate) / max(kalmanRtt, 1) × (1 / (1 + σ/100))

package aethernet.transport

import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write
import kotlin.math.max
import kotlin.math.sqrt

// ── KalmanRttFilter ───────────────────────────────────────────────────────────

/**
 * Two-state Kalman filter estimating RTT and drift for one transport link.
 *
 * State: x = [rtt; drift] — F = [[1,1],[0,1]], H = [1,0].
 *
 * **Not thread-safe.** Callers must hold the selector's write lock before
 * calling [update].
 */
private class KalmanRttFilter(
    initialRttMs:        Double = 200.0,
    private val qRtt:    Double = 25.0,
    private val qDrift:  Double = 5.0,
    private val r:       Double = 100.0,
) {
    // ── State (backing fields are writeable only within this class) ───────
    var rtt:   Double = initialRttMs; private set
    var drift: Double = 0.0;          private set
    var p00:   Double = 400.0;        private set
    var p01:   Double = 0.0;          private set
    var p11:   Double = 100.0;        private set

    /** Posterior variance of the RTT estimate (ms²). Lower = more confident. */
    val rttVariance: Double get() = p00

    /**
     * Incorporate a new RTT measurement and return the updated estimate.
     *
     * Full Kalman predict→update cycle:
     *  1. Predict:  x̂ = F·x,  P̂ = F·P·Fᵀ + Q
     *  2. Gain:     S = H·P̂·Hᵀ + R,  K = P̂·Hᵀ / S
     *  3. Update:   x = x̂ + K·(z − H·x̂),  P = (I − K·H)·P̂
     */
    fun update(measuredRttMs: Double): Double {
        // ── 1. Predict ────────────────────────────────────────────────────────
        val rttPred   = rtt + drift
        val driftPred = drift

        // P_pred = F·P·Fᵀ + Q  (F = [[1,1],[0,1]])
        val pp00 = p00 + 2.0 * p01 + p11 + qRtt
        val pp01 = p01 + p11
        val pp11 = p11 + qDrift

        // ── 2. Kalman gain (H = [1, 0]) ──────────────────────────────────────
        val s  = pp00 + r
        val k0 = pp00 / s
        val k1 = pp01 / s

        // ── 3. Update ─────────────────────────────────────────────────────────
        val innovation = measuredRttMs - rttPred
        rtt   = rttPred   + k0 * innovation
        drift = driftPred + k1 * innovation

        // P = (I − K·H)·P_pred
        p00 = (1.0 - k0) * pp00
        p01 = (1.0 - k0) * pp01
        p11 = -k1 * pp01 + pp11

        // Clamp to prevent numerical drift below zero.
        p00 = maxOf(p00, 1e-6)
        p11 = maxOf(p11, 1e-6)

        return rtt
    }
}

// ── PredictiveTransportSelector ───────────────────────────────────────────────

/**
 * A transport paired with its Kalman-predictive score and uncertainty metadata.
 *
 * @property transport      The ranked transport backend.
 * @property score          Composite predictive score (higher = better).
 * @property predictedRttMs Kalman-estimated RTT in milliseconds.
 * @property rttVariance    Posterior RTT variance (ms²); lower = more confident.
 */
data class RankedTransportPredictive(
    val transport:      TransportService,
    val score:          Double,
    val predictedRttMs: Double,
    val rttVariance:    Double,
)

/**
 * Predictive transport selector maintaining a Kalman RTT filter per transport.
 *
 * Extends [rankTransports] by replacing the EWMA RTT term with a Kalman-estimated
 * RTT (which can predict rising RTT before it crosses a threshold) and adding a
 * reliability penalty proportional to the Kalman variance.
 *
 * Thread-safe via [ReentrantReadWriteLock].
 *
 * ### Usage
 * ```kotlin
 * val selector = PredictiveTransportSelector()
 * transports.forEach { selector.register(it) }
 *
 * // After each send:
 * selector.observeMetrics(transport, rttMs = 42L, success = true, bytesTransferred = 1024L)
 *
 * // Get the best transport for a 500-byte payload:
 * val best = selector.selectBest(payloadBytes = 500)
 * ```
 */
class PredictiveTransportSelector {

    private val lock    = ReentrantReadWriteLock()
    private val filters = mutableMapOf<TransportService, KalmanRttFilter>()

    // ── Registration ──────────────────────────────────────────────────────

    /**
     * Register [transport] for Kalman tracking with an optional initial RTT prior.
     * Safe to call multiple times — subsequent calls for already-registered
     * transports are no-ops.
     */
    fun register(transport: TransportService, initialRttMs: Double = 200.0): Unit =
        lock.write {
            filters.getOrPut(transport) { KalmanRttFilter(initialRttMs) }
        }

    /** Remove [transport] and discard its Kalman state. */
    fun unregister(transport: TransportService): Unit = lock.write {
        filters.remove(transport)
    }

    // ── Observation ───────────────────────────────────────────────────────

    /**
     * Feed a new sample to both the transport's [PerTransportMetrics] EWMA
     * and our Kalman filter.  Call after every completed send attempt.
     *
     * Only successful sends with `rttMs > 0` update the Kalman state —
     * failures carry no useful propagation-delay signal.
     */
    fun observeMetrics(
        transport:        TransportService,
        rttMs:            Long,
        success:          Boolean,
        bytesTransferred: Long,
    ) {
        transport.metrics?.recordSample(rttMs, success, bytesTransferred)

        if (rttMs <= 0L || !success) return

        lock.write {
            filters[transport]?.update(rttMs.toDouble())
        }
    }

    // ── Ranking ───────────────────────────────────────────────────────────

    /**
     * Return transports sorted by predictive score (highest first).
     *
     * Only available transports are included.  [payloadBytes] is used to
     * exclude transports whose max bandwidth would require > 30 s to
     * serialise this payload.
     */
    fun rank(payloadBytes: Int = 512): List<RankedTransportPredictive> = lock.read {
        val result = mutableListOf<RankedTransportPredictive>()

        for ((transport, filter) in filters) {
            if (!transport.isAvailable) continue

            val bw = transport.maxBandwidthBps
            if (bw > 0L) {
                val serialSec = payloadBytes * 8.0 / bw.toDouble()
                if (serialSec > 30.0) continue
            }

            val kalmanRtt = maxOf(filter.rtt, 1.0)
            val variance  = filter.rttVariance
            val stddev    = sqrt(variance)
            val power     = maxOf(transport.powerCostRelative, 1).toDouble()

            val lossRate: Double
            val effectiveBps: Double

            val m = transport.metrics
            if (m != null) {
                lossRate     = m.ewmaLossRate
                effectiveBps = maxOf(m.ewmaThroughputBps, bw * 0.1)
            } else {
                lossRate     = 0.05
                effectiveBps = bw * 0.1
            }

            // Reliability factor: 1.0 at σ=0 ms, ~0.5 at σ=100 ms.
            val reliabilityFactor = 1.0 / (1.0 + stddev / 100.0)
            val score = (effectiveBps / power) * (1.0 - lossRate) / kalmanRtt * reliabilityFactor

            result += RankedTransportPredictive(
                transport      = transport,
                score          = score,
                predictedRttMs = kalmanRtt,
                rttVariance    = variance,
            )
        }

        result.sortedByDescending { it.score }
    }

    /**
     * Return the highest-scoring available transport for [payloadBytes], or `null`.
     */
    fun selectBest(payloadBytes: Int = 512): TransportService? =
        rank(payloadBytes).firstOrNull()?.transport

    /**
     * Return `Triple(rttMs, driftMs, variance)` for a registered transport,
     * or `null` if the transport is not registered.
     */
    fun getKalmanState(transport: TransportService): Triple<Double, Double, Double>? =
        lock.read {
            val f = filters[transport] ?: return@read null
            Triple(f.rtt, f.drift, f.p00)
        }
}
